src/rgw/rgw_rest_s3.cc

```diff
@@ -5870,3 +5875,337 @@ bool rgw::auth::s3::S3AnonymousEngine::is_applicable(
 
   return route == AwsRoute::QUERY_STRING && version == AwsVersion::UNKNOWN;
 }
+
+
+using namespace s3selectEngine;
+const char* RGWSelectObj_ObjStore_S3::header_name_str[3] = {":event-type", ":content-type", ":message-type"};
+const char* RGWSelectObj_ObjStore_S3::header_value_str[3] = {"Records", "application/octet-stream", "event"};
+
+RGWSelectObj_ObjStore_S3::RGWSelectObj_ObjStore_S3():
+  s3select_syntax(new s3select()),
+  m_s3_csv_object(0),
+  m_buff_header(static_cast<char*>(malloc(1000))),
+  chunk_number(0),
+  crc32(0)
+{
+  set_get_data(true);
+}
+
+RGWSelectObj_ObjStore_S3::~RGWSelectObj_ObjStore_S3()
+{
+  if (m_s3_csv_object!=0)
+  {
+    delete m_s3_csv_object;
+  }
+
+  delete s3select_syntax;
+
+  if(m_buff_header)
+  {
+    delete m_buff_header;
```

Comment:
m_buff_header was allocated with malloc(), so needs to be free()d instead of deleted

Reply:
or it could be allocated with new--is there any advantage?

Reply:
for char , i  don't think so.

Reply:
you can just use unique_ptr for all pointers.
no need to change any interface of function using these pointers, just use p.get() to pass them the raw pointers.
if you use unique_ptr, and code never uses "delete" than you know that you memory management is ok.

Reply:
that's true

Reply:
ok,
for the purpose of memory management discussion...
using p.get() (raw pointer ) not undermine the very purpose of uniq_ptr ??
some future change may copy that raw pointer, need to handle that use-case.

Reply:
it doesn't. as long as whoever gets the raw pointer does not delete of free it.
if they do, than they would have an issue with the current code anyway.

if the code that receive the pointer store it somewhere, beyond the original lifetime of the original pointer,  or pass it to another thread, there is another issue.
but again, if this issue exists, it means it already exists today.
and anyway, passing a reference to the unique_ptr won't save you in this case, as this is the case for a shared_ptr.

if you stick to the rule of only using unique_ptr and never using "delete" (or free) in your code, then you memory management is pretty guaranteed.

Reply:
exactly,
current code using bare pointer -- not safe 
in the case your code needs to share pointer , -- for safety use shared_ptr
in the case your code needs single copy for pointer, -- for safety use uniq_ptr
using uniq_ptr.get() is no longer safe from that point. its safe from leakage.

Reply:
shared pointers are a completely different usecase.
bare pointers *are safe* as long as you don't delete them.